### DISCUSSION TITLE
Data binding of child controls of ToolStripControlHost

### DISCUSSION THREAD

**Discussion by hdefucom** [dotnet/winforms#6517] [Q&A] [answered]
* .NET Core Version: 
net40 to net6.0-windows

* Have you experienced this same bug with .NET Framework?: 

Yes

**Problem description:**
Add a ListBox to the ToolStripControlHost.
ListtBox cannot display data bound with datasource,
However, it can be displayed by using items.Add.
This problem only occurs in the ListBox in the ToolStripControlHost.

```

    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();

            _dropDown = new ToolStripDropDown();
            _dropDown.Padding = Padding.Empty;

            box = new ListBox()
            {
                Margin = Padding.Empty,
                Dock = DockStyle.Fill,
                SelectionMode = SelectionMode.MultiSimple
            };

            //*******************************************************  Can display
            //box.Items.AddRange(Enumerable.Range(10000000, 20)
            //    .Select(i => i.ToString())
            //    .Select(i => new TestItem() { Text = i, Value = i })
            //    .OfType<object>().ToArray());
            //*************************************************************************

            //*******************************************************  Cannot display
            box.DisplayMember = "Text";
            box.ValueMember = "Value";
            box.DataSource = new ObservableCollection<TestItem>(
                Enumerable.Range(1, 10).Select(i => i.ToString())
                .Select(i => new TestItem() { Text = i, Value = i })
                );
            //*************************************************************************


            host = new ToolStripControlHost(box) { Padding = Padding.Empty, Margin = Padding.Empty };
            _dropDown.Items.Add(host);
        }

        ListBox box;
        ToolStripControlHost host;
        ToolStripDropDown _dropDown;


        private void button1_Click(object sender, EventArgs e)
        {
            _dropDown.Show(Control.MousePosition);
        }
        private class TestItem
        {
            public string Text { get; set; }
            public string Value { get; set; }

            public override string ToString()
            {
                return Text;
            }
        }
    }



```

**Comment by tbolon** [comment]
We had a similar problem for years in our app, which I think is caused by the same limitation.

The StatusStrip does not supports binding because the method `OnBindingContextChanged` does not update binding on child items. It's caused by the fact that ToolStripItem does not inherit from Control, so is skipped by the base implementation. So ToolStripItem are not aware of any change in parent BindingContext.

We fixed this problem by subclassing StatusStrip and adding a specific method to update the BindingContext on all toolstrip item by overriding OnBindingContextChanged.

For a more general fix, the StatusStrip (and Toolstrip) should override OnBindingContextChanged and update the binding context on all child items, because the base Control.OnBindingContextChanged will not. This code also need to explore all ToolStripItem children.

**Comment by hdefucom** [comment]
> We had a similar problem for years in our app, which I think is caused by the same limitation.
> 
> The StatusStrip does not supports binding because the method `OnBindingContextChanged` does not update binding on child items. It's caused by the fact that ToolStripItem does not inherit from Control, so is skipped by the base implementation. So ToolStripItem are not aware of any change in parent BindingContext.
> 
> We fixed this problem by subclassing StatusStrip and adding a specific method to update the BindingContext on all toolstrip item by overriding OnBindingContextChanged.
> 
> For a more general fix, the StatusStrip (and Toolstrip) should override OnBindingContextChanged and update the binding context on all child items, because the base Control.OnBindingContextChanged will not. This code also need to explore all ToolStripItem children.

How should I rewrite it? I don't know much about the "BindingContext" of the component，Manually call onbindingcontextchanged of toolstripitem?

**Comment by hdefucom** [comment]
> 
This prevents access to protected members, --> CS1540
```
    protected override void OnBindingContextChanged(EventArgs e)
    {
        base.OnBindingContextChanged(e);
        foreach (ToolStripItem item in Items)
        {
            if (item is ToolStripControlHost host)
            {
                host.Control.OnBindingContextChanged(e);
            }
        }
    }
```

**Comment by tbolon** [comment]
It's an old problem and I don't remember exactly the details. We also add to subclass the StatusStripItem if I remember. There is no easy fix.

And by reviewing your code, I see that you are setting directly the DataSource on the control, so I am not even sure my comment applies to your problem. In our case, the DataSource was a BindingSource component attached to the parent form.

You should try creating a new empty BindingContext and attach it to your control, or simply set the BindingContext of the toolstripcontrolhost? In our case it was related to this.

**Comment by hdefucom** [comment]
> It's an old problem and I don't remember exactly the details. We also add to subclass the StatusStripItem if I remember. There is no easy fix.
> 
> And by reviewing your code, I see that you are setting directly the DataSource on the control, so I am not even sure my comment applies to your problem. In our case, the DataSource was a BindingSource component attached to the parent form.
> 
> You should try creating a new empty BindingContext and attach it to your control, or simply set the BindingContext of the toolstripcontrolhost? In our case it was related to this.

Thank you,
I found out how to do it. I just need to manually update the BindingContext after the data source is changed.
But I don't know what consequences this will bring, but it is effective at present.

```
        box.DisplayMember = "Text";
        box.ValueMember = "Value";
        box.DataSource = Enumerable.Range(1, 10).Select(i => i.ToString())
            .Select(i => new TestItem() { Text = i, Value = i }).ToList();
        box.BindingContext = this.BindingContext;

//...............
    private void button2_Click(object sender, EventArgs e)
    {
        box.DataSource.As<List<TestItem>>().Add(new TestItem()
        {
            Text = DateTime.Now.Second.ToString(),
            Value = DateTime.Now.Second.ToString()
        });
        box.BindingContext = null;
        box.BindingContext = this.BindingContext;
    }
```

**Comment by RussKie** [comment, marked as answer]
You need to assign a binding context to the hosted control, otherwise a currency manager object isn't created:
https://github.com/dotnet/winforms/blob/40dd779a6ead7ef755a5fef57d38acc31d0792bb/src/System.Windows.Forms/src/System/Windows/Forms/ListControl.cs#L690-L710

```cs
        box = new ListBox()
        {
            Margin = Padding.Empty,
            Dock = DockStyle.Fill,
            SelectionMode = SelectionMode.MultiSimple
        };

        box.DisplayMember = "Text";
        box.ValueMember = "Value";
        box.DataSource = new ObservableCollection<TestItem>(
            Enumerable.Range(1, 10).Select(i => i.ToString())
            .Select(i => new TestItem() { Text = i, Value = i })
            );

        box.BindingContext = this.BindingContext;    //  <<-- fix

        host = new(box) { Padding = Padding.Empty, Margin = Padding.Empty };
        _dropDown.Items.Add(host);
```

Admittedly I had to step through the code and search the internet myself. In essence here's the answer https://stackoverflow.com/a/3481908/2338036.

Because this is a long standing behaviour changing it carries a significant compat risk, and with such a simple workaround it is unlikely we'll be changing it.
